src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/TempFileTextBufferManager.cs

```diff
@@ -0,0 +1,181 @@
+        }
+
+        public string FilePath { get; set; }
+
+        public Task InitializeBufferAsync() => InitializeAsync();
+
+        protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            FilePath = GetTempFileName(Path.GetFileName(_unconfiguredProject.FullPath));
+
+            var projectXml = await _msbuildAccessor.GetProjectXmlAsync().ConfigureAwait(false);
+            _fileSystem.WriteAllText(FilePath, projectXml, await _unconfiguredProject.GetFileEncodingAsync().ConfigureAwait(false));
+        }
+
+        protected override Task DisposeCoreAsync(bool initialized)
+        {
+            _fileSystem.RemoveDirectory(Path.GetDirectoryName(FilePath), true);
+            return Task.CompletedTask;
+        }
+
+        public async Task ResetBufferAsync()
+        {
+            var projectXml = await _msbuildAccessor.GetProjectXmlAsync().ConfigureAwait(false);
+
+            // We compare the text we want to write with the text currently in the buffer, ignoring whitespace. If they're
+            // the same, then we don't write anything. We ignore whitespace because of
+            // https://github.com/dotnet/roslyn-project-system/issues/743. Once we can read the whitespace correctly from
+            // the msbuild model, we can stop stripping whitespace for this comparison.
+            var normalizedExistingText = _whitespaceRegex.Replace(await ReadBufferXmlAsync().ConfigureAwait(true), "");
+            var normalizedProjectText = _whitespaceRegex.Replace(projectXml, "");
+
+            if (!normalizedExistingText.Equals(normalizedProjectText, StringComparison.Ordinal))
+            {
+                await _threadingService.SwitchToUIThread();
+                // If the docdata is not dirty, we just update the buffer to avoid the file reload pop-up. Otherwise,
+                // we write to disk, to force the pop-up.
+                Verify.HResult(_docData.IsDocDataDirty(out int isDirty));
+                if (Convert.ToBoolean(isDirty))
+                {
+                    _fileSystem.WriteAllText(FilePath, projectXml, await _unconfiguredProject.GetFileEncodingAsync().ConfigureAwait(true));
+                }
+                else
+                {
+                    var textSpan = new Span(0, _textDocument.TextBuffer.CurrentSnapshot.Length);
+
+                    // When the buffer is being reset, it's often been set to ReadOnly. We can't update the buffer when it's readonly, so save
+                    // the currently flags and turn off ReadOnly, restoring after we're done. We're on the UI thread, so this is invisible to
+                    // the user.
+                    Verify.HResult(_textBuffer.GetStateFlags(out uint oldFlags));
+                    _textBuffer.SetStateFlags(oldFlags & ~(uint)BUFFERSTATEFLAGS.BSF_USER_READONLY);
+
+                    _textDocument.TextBuffer.Replace(textSpan, projectXml);
```

Comment:
Does this throw? Because you're going to be leaving this as read-only if it does.